open-machine-learning-jupyter-book/deep-learning/DQN.md

```diff
@@ -13,10 +13,367 @@ kernelspec:
+        return len(self.buffer)
+
+class ActionStateModel:
+    def __init__(self, state_dim, aciton_dim):
+        self.state_dim  = state_dim
+        self.action_dim = aciton_dim
+        self.epsilon = args.eps
+        
+        self.model = self.create_model()
+    
+    def create_model(self):
+        model = tf.keras.Sequential([
+            Input((self.state_dim,)),
+            Dense(32, activation='relu'),
+            Dense(16, activation='relu'),
+            Dense(self.action_dim)
+        ])
+        model.compile(loss='mse', optimizer=Adam(args.lr))
+        return model
+    
+    def predict(self, state):
+        return self.model.predict(state)
+    
+    def get_action(self, state):
+        state = np.reshape(state, [1, self.state_dim])
+        self.epsilon *= args.eps_decay
+        self.epsilon = max(self.epsilon, args.eps_min)
+        q_value = self.predict(state)[0]
+        if np.random.random() < self.epsilon:
+            return random.randint(0, self.action_dim-1)
+        return np.argmax(q_value)
+
+    def train(self, states, targets):
+        self.model.fit(states, targets, epochs=1, verbose=0)
+    
+
+class Agent:
+    def __init__(self, env):
+        self.env = env
+        self.state_dim = self.env.observation_space.shape[0]
+        self.action_dim = self.env.action_space.n
+
+        self.model = ActionStateModel(self.state_dim, self.action_dim)
+        self.target_model = ActionStateModel(self.state_dim, self.action_dim)
+        self.target_update()
+
+        self.buffer = ReplayBuffer()
+
+    def target_update(self):
+        weights = self.model.model.get_weights()
+        self.target_model.model.set_weights(weights)
+    
+    def replay(self):
+        for _ in range(10):
+            states, actions, rewards, next_states, done = self.buffer.sample()
+            targets = self.target_model.predict(states)
+            next_q_values = self.target_model.predict(next_states).max(axis=1)
+            targets[range(args.batch_size), actions] = rewards + (1-done) * next_q_values * args.gamma
+            self.model.train(states, targets)
+    
+    def train(self, max_episodes=1000):
+        for ep in range(max_episodes):
+            done, total_reward = False, 0
+            state = self.env.reset()
+            while not done:
+                action = self.model.get_action(state)
+                next_state, reward, done, _ = self.env.step(action)
+                self.buffer.put(state, action, reward*0.01, next_state, done)
+                total_reward += reward
+                state = next_state
+            if self.buffer.size() >= args.batch_size:
+                self.replay()
+            self.target_update()
+            print('EP{} EpisodeReward={}'.format(ep, total_reward))
+            wandb.log({'Reward': total_reward})
+
+
+def main():
+    env = gym.make('CartPole-v1')
+    agent = Agent(env)
+    agent.train(max_episodes=1000)
+
+if __name__ == "__main__":
+    main()

```

Comment:
The code here and the text example is not the same game?

Reply:
Yes, I changed the code part in the maze game.